src/main/java/io/cryostat/agent/WebServer.java

```diff
@@ -164,6 +171,59 @@ public String description() {
         }
     }
 
+    private class CompressionFilter extends Filter {
+
+        @Override
+        public void doFilter(HttpExchange exchange, Chain chain) throws IOException {
+            List<String> encodings = new ArrayList<>();
+            List<String> rawEncodings =
+                    exchange.getRequestHeaders().getOrDefault("Accept-Encoding", List.of());
+            rawEncodings.forEach(
+                    raw -> {
+                        if (raw.contains(",")) {
+                            encodings.addAll(Arrays.asList(raw.replaceAll("\\s", "").split(",")));
+                        } else {
+                            encodings.add(raw);
+                        }
+                    });
+            String actualEncoding = null;
+            priority:
+            for (String requestedEncoding : encodings) {
+                switch (requestedEncoding) {
+                    case "deflate":
+                        actualEncoding = requestedEncoding;
+                        exchange.setStreams(
+                                exchange.getRequestBody(),
+                                new DeflaterOutputStream(exchange.getResponseBody()));
+                        break priority;
+                        // TODO gzip encoding breaks communication with the server, need to
```

Comment:
The Cryostat server supports `gzip` and `deflate` (and maybe more) via Vert.x, but the -agent instance's built-in webserver only implements `deflate` right now. If you tried that same experiment but with a request URL directed at the -agent instance, instead of to the Cryostat server acting as a proxy, then you would receive back an unencoded response since the negotiation would fail on selecting `gzip` and just send back the raw bytestream.

Reply:
Basically from `http` to Cryostat you have a request for `gzip` responses if possible, and Cryostat respects that. But when Cryostat talks to the -agent it asks for `gzip`, otherwise `deflate`, and the -agent says "sure, here's a `deflate` response." Cryostat receives that data and decompresses it, and recompresses it `gzip` since that's what `http` asked for.

Reply:
Oh, I see right. I would have to query the -agent itself with that header to see the results. Makes sense, thanks!